File: portal_ibr_backend/services/tarefas_service.py
from datetime import datetime

from repositories.tarefas_repository import TarefasRepository
from schemas.tarefa_schema import TarefaSchema
from utils.responses import success, error
from utils.exceptions import AppError
from utils.logger import logger


STATUS_A_FAZER = "A Fazer"
STATUS_FAZENDO = "Fazendo"
STATUS_CONCLUIDO = "Concluído"


def serializar_tarefa(tarefa):
    return {
        "_id": str(tarefa["_id"]),

        "ministerio":
            tarefa.get(
                "ministerio",
                ""
            ),

        "solicitante":
            tarefa.get(
                "solicitante",
                ""
            ),

        "formatos_solicitados":
            tarefa.get(
                "formatos_solicitados",
                []
            ),

        "descricao":
            tarefa.get(
                "descricao",
                ""
            ),

        "sugestao_arte":
            tarefa.get(
                "sugestao_arte",
                ""
            ),

        "data_evento":
            tarefa.get(
                "data_evento",
                ""
            ),

        "horario_evento":
            tarefa.get(
                "horario_evento",
                ""
            ),

        "data_entrega":
            tarefa.get(
                "data_entrega",
                ""
            ),

        "status":
            tarefa.get(
                "status",
                STATUS_A_FAZER
            ),

        "responsavel_id":
            tarefa.get(
                "responsavel_id"
            ),

        "responsavel_nome":
            tarefa.get(
                "responsavel_nome"
            ),

        "data_cadastro": (
            tarefa.get(
                "data_cadastro"
            ).isoformat()
            if tarefa.get(
                "data_cadastro"
            )
            else None
        )
    }


def listar_tarefas_service():
    try:
        tarefas = [
            serializar_tarefa(
                tarefa
            )
            for tarefa
            in TarefasRepository.listar()
        ]

        return success(
            data=tarefas,
            total=len(tarefas)
        )

    except Exception as e:
        logger.exception(
            "Erro ao listar tarefas"
        )

        return error(
            str(e),
            500
        )


def buscar_tarefa_service(id):
    try:
        tarefa = (
            TarefasRepository.buscar_por_id(
                id
            )
        )

        if not tarefa:
            return error(
                "Tarefa não encontrada.",
                404
            )

        return success(
            data=serializar_tarefa(
                tarefa
            )
        )

    except Exception as e:
        logger.exception(
            "Erro ao buscar tarefa"
        )

        return error(
            str(e),
            500
        )


def cadastrar_tarefa_service(data):
    try:
        dados = TarefaSchema.validar(
            data
        )

        documento = {
            **dados,

            "status":
                STATUS_A_FAZER,

            "responsavel_id":
                None,

            "responsavel_nome":
                None,

            "data_cadastro":
                datetime.utcnow()
        }

        resultado = (
            TarefasRepository.cadastrar(
                documento
            )
        )

        return success(
            data={
                "id":
                    str(
                        resultado.inserted_id
                    )
            },
            message=(
                "Tarefa cadastrada "
                "com sucesso."
            ),
            status=201
        )

    except AppError as e:
        return error(
            e.message,
            e.status
        )

    except Exception as e:
        logger.exception(
            "Erro ao cadastrar tarefa"
        )

        return error(
            str(e),
            500
        )


def editar_tarefa_service(
    id,
    data
):
    try:
        tarefa = (
            TarefasRepository.buscar_por_id(
                id
            )
        )

        if not tarefa:
            return error(
                "Tarefa não encontrada.",
                404
            )

        dados = TarefaSchema.validar(
            data
        )

        TarefasRepository.atualizar(
            id,
            dados
        )

        return success(
            message=(
                "Tarefa atualizada "
                "com sucesso."
            )
        )

    except AppError as e:
        return error(
            e.message,
            e.status
        )

    except Exception as e:
        logger.exception(
            "Erro ao editar tarefa"
        )

        return error(
            str(e),
            500
        )


def excluir_tarefa_service(id):
    try:
        tarefa = (
            TarefasRepository.buscar_por_id(
                id
            )
        )

        if not tarefa:
            return error(
                "Tarefa não encontrada.",
                404
            )

        TarefasRepository.excluir(
            id
        )

        return success(
            message=(
                "Tarefa excluída "
                "com sucesso."
            )
        )

    except Exception as e:
        logger.exception(
            "Erro ao excluir tarefa"
        )

        return error(
            str(e),
            500
        )


def assumir_tarefa_service(
    id,
    usuario
):
    try:
        tarefa = (
            TarefasRepository.buscar_por_id(
                id
            )
        )

        if not tarefa:
            return error(
                "Tarefa não encontrada.",
                404
            )

        if (
            tarefa.get("status")
            != STATUS_A_FAZER
        ):
            return error(
                "Esta tarefa não está "
                "disponível para ser assumida.",
                400
            )

        integrante_id = usuario.get(
            "id"
        )

        integrante_nome = usuario.get(
            "nome"
        )

        if (
            not integrante_id
            or not integrante_nome
        ):
            return error(
                "Não foi possível identificar "
                "o integrante logado.",
                400
            )

        dados = {
            "status":
                STATUS_FAZENDO,

            "responsavel_id":
                integrante_id,

            "responsavel_nome":
                integrante_nome
        }

        TarefasRepository.atualizar(
            id,
            dados
        )

        return success(
            message=(
                "Tarefa assumida "
                "com sucesso."
            )
        )

    except Exception as e:
        logger.exception(
            "Erro ao assumir tarefa"
        )

        return error(
            str(e),
            500
        )


def concluir_tarefa_service(
    id,
    usuario
):
    try:
        tarefa = (
            TarefasRepository.buscar_por_id(
                id
            )
        )

        if not tarefa:
            return error(
                "Tarefa não encontrada.",
                404
            )

        if (
            tarefa.get("status")
            != STATUS_FAZENDO
        ):
            return error(
                "Esta tarefa não está "
                "em andamento.",
                400
            )

        integrante_id = usuario.get(
            "id"
        )

        if (
            tarefa.get(
                "responsavel_id"
            )
            != integrante_id
        ):
            return error(
                "Somente o responsável "
                "pela tarefa pode concluí-la.",
                403
            )

        TarefasRepository.atualizar(
            id,
            {
                "status":
                    STATUS_CONCLUIDO
            }
        )

        return success(
            message=(
                "Tarefa concluída "
                "com sucesso."
            )
        )

    except Exception as e:
        logger.exception(
            "Erro ao concluir tarefa"
        )

        return error(
            str(e),
            500
        )

File: portal_ibr_frontend/src/app/core/services/tarefas.service.ts
import {
  Injectable,
  inject
} from '@angular/core';

import {
  HttpClient
} from '@angular/common/http';

import {
  Observable
} from 'rxjs';

import {
  environment
} from '../../../environments/environment';


export type StatusTarefa =
  | 'A Fazer'
  | 'Fazendo'
  | 'Concluído';


export interface Tarefa {
  _id: string;

  ministerio: string;

  solicitante: string;

  formatos_solicitados: string[];

  descricao: string;

  sugestao_arte: string;

  data_evento: string;

  horario_evento: string;

  data_entrega: string;

  status: StatusTarefa;

  responsavel_id: string | null;

  responsavel_nome: string | null;

  data_cadastro: string | null;
}


export interface TarefasResponse {
  success: boolean;

  data: Tarefa[];

  total: number;

  message?: string;
}


export interface TarefaResponse {
  success: boolean;

  data: Tarefa;

  message?: string;
}


export interface SalvarTarefaRequest {
  ministerio: string;

  solicitante: string;

  formatos_solicitados: string[];

  descricao: string;

  sugestao_arte: string;

  data_evento: string;

  horario_evento: string;

  data_entrega: string;
}


export interface AcaoTarefaResponse {
  success: boolean;

  data?: {
    id?: string;
  };

  message?: string;
}


@Injectable({
  providedIn: 'root'
})
export class TarefasService {

  private readonly http =
    inject(HttpClient);

  private readonly apiUrl =
    `${environment.apiUrl}/api/tarefas`;


  listar(): Observable<TarefasResponse> {

    return this.http.get<TarefasResponse>(
      this.apiUrl
    );

  }


  buscarPorId(
    tarefaId: string
  ): Observable<TarefaResponse> {

    return this.http.get<TarefaResponse>(
      `${this.apiUrl}/${tarefaId}`
    );

  }


  cadastrar(
    dados: SalvarTarefaRequest
  ): Observable<AcaoTarefaResponse> {

    return this.http.post<AcaoTarefaResponse>(
      this.apiUrl,
      dados
    );

  }


  editar(
    tarefaId: string,
    dados: SalvarTarefaRequest
  ): Observable<AcaoTarefaResponse> {

    return this.http.put<AcaoTarefaResponse>(
      `${this.apiUrl}/${tarefaId}`,
      dados
    );

  }


  excluir(
    tarefaId: string
  ): Observable<AcaoTarefaResponse> {

    return this.http.delete<AcaoTarefaResponse>(
      `${this.apiUrl}/${tarefaId}`
    );

  }


  assumir(
    tarefaId: string
  ): Observable<AcaoTarefaResponse> {

    return this.http.put<AcaoTarefaResponse>(
      `${this.apiUrl}/${tarefaId}/assumir`,
      {}
    );

  }


  concluir(
    tarefaId: string
  ): Observable<AcaoTarefaResponse> {

    return this.http.put<AcaoTarefaResponse>(
      `${this.apiUrl}/${tarefaId}/concluir`,
      {}
    );

  }

}